File: old/session02Sorting/Quicksort.java
public class Quicksort {
    public static void quicksort(int[] x, int L, int R) {
	if ( L >= R)
	    return;

	int i = L, j = R;

	// 	partition into two halves
	while (i < j) {
	    while (x[i] <= pivot)
		i++;
	    while (x[j] > pivot)
		j--;
	    swap(x[i], x[j]);
	}

	// i == j somewhere in the middle
	quicksort(x, L, i);
	quicksort(x, i+1, R);


    }

}

File: old/session02Sorting/quicksort.cc
 0//quicksort
/*
8  7  4  5  6  3  1  2
pivot = 4
         i     j
2                    8
   1              7
         3     5
2  1  4  3  6  5  7  8
         j   i

        O(n log n)
*/
void quicksort(int a[], int L, int R)
{
    // termination condition
    if (L >= R)
        return;

    int pivot = a[random(L, R)];
    int i = L, j = R;
    while (i < j)
    {
        while (i < j && a[i] <= pivot)
            i++;
        while (i < j && a[j] > pivot)
            j--;
        // what is true here?
        if (i < j)
        {
            swap(a[i], a[j]);
            i++, j--;
        }
    }
    quicksort(a, L, i - 1);
    quicksort(a, i, R);
}

//LOMUTO